[Version 3.6] Add GenerateCoverageViz pipeline step to dag.

MarkAZhang opened the pull request:
Also slight change to BlastContigs pipeline step to generate blast.top.m8 file.

Testing steps:
* Tested the benchmark sample with this dag branch on staging, and it succeeded as expected. https://staging.idseq.net/samples/12576
* Tested a minimal sample with no assembled contigs with this dag branch on staging, and it also succeeded (this tests the case where some files are empty) https://staging.idseq.net/samples/12575

The coverage viz displays correctly using the files generated by this pipeline step.

Step takes about 1 minute to run for the benchmark sample, based on cloudwatch logs.

yunfangjuan commented on examples/assemble_blast.json:
```diff
@@ -28,13 +28,15 @@
     "rapsearch2_accessions_out": ["nr.refseq.fasta"],
     "refined_gsnap_out": [
       "gsnap.blast.m8",
+      "gsnap.blast.top.m8",
```
move this to the end of the json?

cdebourcy replied:
Agreed, it might be less prone to human error if new outputs are added to the end of the list rather than in the middle. Or just confirm that all the steps that refer to `refined_gsnap_out` still refer to the correct indices of these lists. For example, the postprocess dag in idseq-web (https://github.com/chanzuckerberg/idseq-web/blob/7da56caff3a686bcf2559c51ece28a0f258779c5/app/lib/dags/postprocess.json.erb#L146) indicates that the step combine_taxon_counts (https://github.com/chanzuckerberg/idseq-dag/blob/master/idseq_dag/steps/combine_taxon_counts.py#L12) is pulling the item at index 3 of the `refined_gsnap_out` list as an input. But your change shifts the indices of this list, so it would break

cdebourcy replied:
[Note to self for future: It might be a good idea to change the targets into dicts so we can use named keys and stop referring to target items by their index in the code for the pipeline steps. Might speed up development]

yunfangjuan commented on idseq_dag/steps/blast_contigs.py:
```diff
@@ -30,7 +30,7 @@ def run(self):
         reference_fasta = self.input_files_local[2][0]
 
-        (blast_m8, refined_m8, refined_hit_summary, refined_counts, contig_summary_json) = self.output_files_local()
+        (blast_m8, top_entry_m8, refined_m8, refined_hit_summary, refined_counts, contig_summary_json) = self.output_files_local()
```
top entry to the end of the output?

yunfangjuan commented on idseq_dag/steps/download_accessions.py:
```diff
@@ -122,5 +149,21 @@ def fetch_sequence_for_thread(error_flags, accession, accession_out_file,
+                entry = loc_dict.get(accession_id)
+                if entry and len(entry) == 3:
+                    # The seq_len includes newline characters in the fasta file, so we must correct for this.
+                    seq_len = math.floor(FASTA_SEQ_LINE_LENGTH / (FASTA_SEQ_LINE_LENGTH + 1) * int(entry[2]))
```
hm it's still not clear what this is.  also I think it's a bit dangerous to assume FASTA_SEQ_LINE_LENGTH is 80. Can you just make this a rough length?

MarkAZhang replied:
This calculates the number of nucleotide bases, and removes the newline characters from the count.

Good point about assuming 80. I will see if I can calculate dynamically in some way.

cdebourcy replied:
If I understand correctly, the issue is that we cannot use the size of the byterange as the sequence length, because byterange includes new-line characters. I can see a few different options, some of which are much better than others:

(a) Calculate all the sequence lengths from the downloaded sequence files. This would be super wasteful to do at pipeline run time.
(b) Determine the actual sequence length while generating nt_loc_db (this is part of the index update operation) and store it in the dict once and for all so it can just be looked up. Probably the best option, but requires the development/testing overhead of building a new nt_loc_db.
(c) Keep the current setup with assumed line length of 80, but do a quick spot check in the code to confirm sequence files actually satisfy this condition. This code can probably live in the step that generates a new nt_loc_db when we update the index. Include an assertion with a very explicit / verbose message for the developers if the condition fails, so that they know how to address the problem (for example, "you need to change the value of FASTA_SEQ_LINE_LENGTH in download_accessions.py"). For what it's worth, I do think the line length of 80 will be a good assumption for a long time to come. Wikipedia article says it's something of a standard: https://en.wikipedia.org/wiki/FASTA_format

Anything I missed?

yunfangjuan commented on idseq_dag/steps/download_accessions.py:
```diff
@@ -79,14 +94,26 @@ def download_ref_sequences_from_s3(self, accession_dict, output_reference_fasta,
+            cmd = f"cat {accession_out_file} |head -1"
+            seq_name = subprocess.check_output(
+                    cmd, executable='/bin/bash', shell=True).decode("utf-8").split(" ", 1)[1].replace("\n", "")
+            accession_summary[accession]["name"] = seq_name
```
this might actually not be thread-safe though. Also, it's probably easier to get the sequence length here?

MarkAZhang replied:
What would make something not thread-safe? Not sure what I should be thinking about.

cdebourcy replied:
@yunfangjuan do you want to elaborate? This happens after the threads are joined, no?

yunfangjuan replied:
Writing to the same dict simultaneously

cdebourcy replied:
To me it looks as if all the download threads are joined in line 94 (https://github.com/chanzuckerberg/idseq-dag/blob/ffef5bf9a6aa7b0284775d9f11a7818e28dc4577/idseq_dag/steps/download_accessions.py#L94) and writing to the dict is single-threaded.

The two places in the code that write to `accession_summary`, namely the methods `download_ref_sequences_from_s3` and `get_sequence_by_accession_from_file`, are called in a mutually exclusive way in the code.

@yunfangjuan Anything I'm missing? Please advise.

yunfangjuan replied:
OK. This case it's actually done after the join so there's only one thread doing it, so it's fine. But I think a better way to deal with this will be to generate another dictionary from the generate_loc_db step to pull out the seq_name and sequence length once for each index update so you won't add time to our current pipeline.

yunfangjuan commented on idseq_dag/steps/download_accessions.py:
```diff
@@ -122,5 +149,21 @@ def fetch_sequence_for_thread(error_flags, accession, accession_out_file,
         finally:
             semaphore.release()
 
+    def generate_dummy_fasta_ref(self, output_reference_fasta):
```
what's this function for?

MarkAZhang replied:
WIP placeholder function.

yunfangjuan commented:
I think b) is good and is actually pretty easy to do. You change the step,
calculate length and output it to a new sqlite3 file. It can also serve
other purposes if people are trying to curate NT/NR.

On Wed, Apr 10, 2019 at 1:20 PM Charles de Bourcy <notifications@github.com>
wrote:

> *@cdebourcy* commented on this pull request.
> ------------------------------
>
> In idseq_dag/steps/download_accessions.py
> <https://github.com/chanzuckerberg/idseq-dag/pull/133#discussion_r274139349>
> :
>
> >
> +        with open(accession_summary_tab, 'w') as ast:
> +            for accession_id, _ in accession_dict.items():
> +                entry = loc_dict.get(accession_id)
> +                if entry and len(entry) == 3:
> +                    # The seq_len includes newline characters in the fasta file, so we must correct for this.
> +                    seq_len = math.floor(FASTA_SEQ_LINE_LENGTH / (FASTA_SEQ_LINE_LENGTH + 1) * int(entry[2]))
>
> If I understand correctly, the issue is that we cannot use the size of the
> byterange as the sequence length, because byterange includes new-line
> characters. I can see a few different options, some of which are much
> better than others:
>
> (a) Calculate all the sequence lengths from the downloaded sequence files.
> This would be super wasteful to do at pipeline run time.
> (b) Determine the actual sequence length while generating nt_loc_db (this
> is part of the index update operation) and store it in the dict once and
> for all so it can just be looked up. Probably the best option, but requires
> the development/testing overhead of building a new nt_loc_db.
> (c) Keep the current setup with assumed line length of 80, but do a quick
> spot check in the code to confirm sequence files actually satisfy this
> condition. This code can probably live in the step that generates a new
> nt_loc_db when we update the index. Include an assertion with a very
> explicit / verbose message for the developers if the condition fails, so
> that they know how to address the problem (for example, "you need to change
> the value of FASTA_SEQ_LINE_LENGTH in download_accessions.py"). For what
> it's worth, I do think the line length of 80 will be a good assumption for
> a long time to come. Wikipedia article says it's something of a standard:
> https://en.wikipedia.org/wiki/FASTA_format
>
> Anything I missed?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/chanzuckerberg/idseq-dag/pull/133#discussion_r274139349>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABoAQnVNdCYV_wKpD8Jxf7H2Oq5T37N2ks5vfkdwgaJpZM4cWpDr>
> .
>

cdebourcy commented on tests/generate_coverage_viz.py:
```diff
@@ -0,0 +1,344 @@
+
+from idseq_dag.steps.generate_coverage_viz import PipelineStepGenerateCoverageViz
+
+class GenerateCoverageViz(unittest.TestCase):
```
Love these tests!

yunfangjuan commented on idseq_dag/steps/generate_coverage_viz.py:
```diff
@@ -0,0 +1,241 @@
+        return reads
+
+
+    def augment_contigs_map_with_coverage(self, contig_coverage_json, contigs_map):
```
Do you need self? Can make this s staticmethod?

yunfangjuan commented:
In general, this is rad!!

cdebourcy commented on idseq_dag/steps/generate_coverage_viz.py:
```diff
@@ -0,0 +1,241 @@
+            with open(upload_file, 'w') as uf:
+                json.dump(coverage_viz_obj[accession_id], uf)
+
+            self.additional_files_to_upload.append(upload_file)
```
This needs to be changed to use `additional_folders_to_upload` instead of `additional_files_to_upload`. The former is a mechanism I just introduced in my last PR. It uploads everything in the folder to S3 in a single AWS CLI request. The latter uploads every file using an individual request. For thousands and thousands of files it's going to take hours.

cdebourcy replied:
Please review the optimizations I introduced to cut the experimental stage from 3 days to 3 hours here: https://github.com/chanzuckerberg/idseq-dag/pull/139

cdebourcy commented on idseq_dag/steps/generate_coverage_viz.py:
```diff
@@ -0,0 +1,241 @@
+                    contigs_map[contig_name]["coverage"] = contig_coverage[contig_name]["coverage"]
+
+
+    def augment_contigs_map_with_byteranges(self, contigs_fasta, contigs_map):
```
Yay!

cdebourcy commented on idseq_dag/steps/generate_coverage_viz.py:
```diff
@@ -0,0 +1,241 @@
+
+            if parts[0] in unassigned_reads_set:
+                reads[parts[0]] = {
+                    # Extract the length from the contig name.
```
Why the code duplication? Can make a helper to parse the general m8 into this structure

cdebourcy commented on idseq_dag/steps/generate_coverage_viz.py:
```diff
@@ -0,0 +1,241 @@
+            if parts[0] in valid_contigs_with_read_counts:
+                contigs[parts[0]] = {
+                    "total_length": int(name_parts[3]),
+                    "accession": parts[1],
```
duplicated with line 199

cdebourcy commented:
Also, could you please test a few large, real-life samples? In order to assess performance and results for different distributions of reads across contigs and accessions.

The benchmark has a low number of reads and was generated from a small number of species covered really really well which makes it easy to assemble them into small numbers of contigs. Real samples will have a much much longer tail of singletons

MarkAZhang commented:
Regarding utils vs static methods, it's pretty arbitrary.

The static methods are mostly reading files, and the util methods are calculations and JSON object compilation.

One other distinction is the util methods are all tested with unit tests, and the static methods aren't since I wasn't sure how to mock files.

MarkAZhang commented:
Talked to @yunfangjuan today about this PR.

She had two concerns:
* I wasn't taking paired reads into account, and she remembered some tricky logic in alignment-viz and `ReadViz.jsx` involving paired reads. I dug into this, and it turns out paired reads can affect the indices for the read, but not for the reference sequence. Since I only need the indices for the reference sequence to figure out where to add coverage to account for the read, I don't need to take paired reads into account for coverage viz. (in alignment viz, we also need to extract the exact aligned sequence from the read, in the correct orientation, which is why it's important)
* She was concerned about my putting a lot of functions into the `utils` folder that were specific to the CoverageViz step. She was also concerned that my code was a bit hard to read because I didn't group my functions together well enough. I'll do a pass to package/modularize my functions into neater modules with clear inputs/outputs.

MarkAZhang commented:
Latest commit:
Improve code organization for coverage viz pipeline step.
    
 Removed util/coverage and moved functions into PipelineStepGenerateCoverageViz,
 since those functions are specific to that one pipeline step.
    
 Reorganized functions in PipelineStepGenerateCoverageViz so that the "main"
 functions are at the top of the file.
  
 Consolidated tests into one file.

Verified that:
* Tests still pass.
* The pipeline step still outputs the exact same JSON as before.

yunfangjuan commented on idseq_dag/steps/generate_coverage_viz.py:
```diff
@@ -0,0 +1,823 @@
+
+        return valid_contigs_with_read_counts
+
+    # Generate a dict that maps accessions to the reads and contigs that were assigned to them.
```
I believe for the python convention you will put the documentation inside the function with ''' ''' around it.

MarkAZhang replied:
Done.

yunfangjuan commented on idseq_dag/steps/generate_coverage_viz.py:
```diff
@@ -0,0 +1,823 @@
+# Decrement the lower bound of a possibly inverted interval.
+# i.e. turn (1, 5) into (0, 5) and (5, 1) into (5, 0).
+# Intervals in m8 alignment files can be inverted.
+def _decrement_lower_bound(interval):
```
I like it.

MarkAZhang commented on idseq_dag/steps/generate_coverage_viz.py:
```diff
@@ -0,0 +1,246 @@
+MIN_M8_FILE_SIZE = 25 # M8 file should have at least a single line.
+
+# These can be overridden with additional_attributes
+MAX_NUM_BINS_COVERAGE = 500
```
Added comment.
